controller/backup_controller.go

```diff
@@ -0,0 +1,444 @@
+		log.Debugf("Backup snapshot initialized")
+
+		_, err = engine.SnapshotBackup(
+			backup.Name, backup.Spec.SnapshotName, backupTarget,
+			backup.Spec.BackingImage, backup.Spec.BackingImageURL,
+			backup.Spec.Labels, credential)
+		if err != nil {
+			bc.eventRecorder.Eventf(backup, corev1.EventTypeWarning, EventReasonInitialized,
+				"Error volume %s snapshot %s label %v: %v",
+				volumeName, backup.Spec.SnapshotName, backup.Spec.Labels, err)
+			log.WithError(err).Errorf("Failed to backup snapshot initialized")
+			return
+		}
+
+		bc.eventRecorder.Eventf(backup, corev1.EventTypeNormal, EventReasonInProgress,
+			"Volume %s snapshot %s label %v", volumeName, backup.Spec.SnapshotName, backup.Spec.Labels)
+		log.Debugf("Backup snapshot in progress")
+
+		bks := &types.BackupStatus{}
+		for {
+			engines, err := bc.ds.ListVolumeEngines(volumeName)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to get engines for volume %v", volumeName)
+				bc.eventRecorder.Eventf(backup, corev1.EventTypeWarning, EventReasonFailed,
+					"Get engines for volume %s error %v", volumeName, err)
+				return
+			}
+
+			for _, e := range engines {
+				backupStatusList := e.Status.BackupStatus
+				for _, b := range backupStatusList {
+					if b.SnapshotName == backup.Spec.SnapshotName {
+						bks = b
+						break
+					}
+				}
+			}
+			if bks.Error != "" {
+				bc.eventRecorder.Eventf(backup, corev1.EventTypeWarning, EventReasonFailed,
+					"Backup error %v volume %s snapshot %s label %v", bks.Error, volumeName, backup.Spec.SnapshotName, backup.Spec.Labels)
+				log.Errorf("Failed to updated volume LastBackup for %s due to backup error %s", volumeName, bks.Error)
+				break
+			}
+			if bks.Progress == 100 {
+				// Request backup volume controller to reconcile BackupVolume immediately.
+				backupVolume, err := bc.ds.GetBackupVolume(volumeName)
+				if err == nil {
+					backupVolume.Spec.SyncRequestAt = &metav1.Time{Time: time.Now().Add(time.Second).UTC()}
+					if _, err = bc.ds.UpdateBackupVolume(backupVolume); err != nil && !datastore.ErrorIsConflict(err) {
+						log.WithError(err).Errorf("Error updating backup volume %s spec", volumeName)
+					}
+				} else if err != nil && datastore.ErrorIsNotFound(err) {
+					backupVolume := &longhorn.BackupVolume{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: volumeName,
+						},
+						Spec: types.BackupVolumeSpec{
+							SyncRequestAt: &metav1.Time{Time: time.Now().Add(time.Second).UTC()},
+						},
+					}
+					if _, err = bc.ds.CreateBackupVolume(backupVolume); err != nil {
+						log.WithError(err).Errorf("Error creating backup volume %s into cluster", volumeName)
+					}
+				}
+
+				bc.eventRecorder.Eventf(backup, corev1.EventTypeNormal, EventReasonSucceeded,
+					"Volume %s snapshot %s label %v", volumeName, backup.Spec.SnapshotName, backup.Spec.Labels)
+				log.Debugf("Backup snapshot succeeded")
+				break
```

Comment:
I am thinking about how the backup volume config file gets created,. If it's created automatically when the 1st backup is ready. We can leave the backup target controller to pull it and create the CR. Then the backup target controller is the only owner of the resource. (only one that is responsible for managing BackupVolume CR lifecycle.)

Reply:
I need to double-check it again. But If I remember correctly, the backup volume config creates once the snapshot backup is started.